skimage/transform/integral.py

```diff
@@ -27,7 +27,13 @@ def integral_image(image):
            ACM SIGGRAPH Computer Graphics, vol. 18, 1984, pp. 207-212.
 
     """
-    S = image
+    if image.real.dtype.kind == 'f':
```

Comment:
I am not sure to understand this change. Is it a bug fix? Otherwise it is a breaking change :confused:

Reply:
It can be fine to keep the precision of the input if the image is not too large, but these cumulative sums can become very inaccurate for large image sizes. The basic issue is that the cumulative sum may be very large and then you are doing addition of relatively small values to very large values which leads to inaccuracy. The same effect is still present in float64, but just won't be as noticeable until larger image sizes are reached. 

There was a concrete example using the integral images implemented internally to non-local means ([see here](https://github.com/scikit-image/scikit-image/pull/2878#issuecomment-346344773) and in the subsequent comment). In that case the error is highly visible in the image for float32 when the input is shape 600x1000x1000, but not visually noticeable for float64.

One compromise to avoid a behavior change would be to do the cumsum in float64, but then cast the final sum to the dtype of the input?

Reply:
I have reverted this dtype change for integral_image and will raise a separate PR related to that instead